Fix all typing issues

Kludex opened the pull request:
@fantix can you please check this?

Kludex committed:
Fix all typing issues

Kludex committed:
push

Kludex committed:
less permissive

Kludex commented on httptools/parser/parser.pyi:
```diff
@@ -1,57 +1,59 @@
-        protocol -- a Python object with the following methods
-        (all optional):
-
-          - on_message_begin()
-          - on_url(url: bytes)
-          - on_header(name: bytes, value: bytes)
-          - on_headers_complete()
-          - on_body(body: bytes)
-          - on_message_complete()
-          - on_chunk_header()
-          - on_chunk_complete()
-          - on_status(status: bytes)
```
This is not needed. Implementers can follow the protocol.

Kludex commented on httptools/parser/parser.pyi:
```diff
@@ -1,57 +1,59 @@
-          - on_chunk_header()
-          - on_chunk_complete()
-          - on_status(status: bytes)
+    def __init__(self, protocol: HTTPProtocol | object) -> None:
```
`object` is less permissive than `Any`.

---

Python still doesn't support "not required class method protocol": https://discuss.python.org/t/discussion-optional-class-and-protocol-fields-and-methods

Kludex commented on httptools/parser/parser.pyi:
```diff
@@ -1,57 +1,59 @@
+            protocol (HTTPProtocol): Callback interface for the parser.
         """
 
+    def set_dangerous_leniencies(
```
This was missing.

Kludex commented on httptools/parser/parser.pyi:
```diff
@@ -1,57 +1,59 @@
-        ...
 
-    def feed_data(self, data: Union[bytes, bytearray, memoryview, array]) -> None:
+    def feed_data(self, data: Union[bytes, bytearray, memoryview, array[int]]) -> None:
```
`array` is a `Generic`.

Kludex commented on httptools/parser/protocol.py:
```diff
@@ -4,12 +4,12 @@
-    def on_chunk_header() -> None: ...
-    def on_chunk_complete() -> None: ...
-    def on_status(status: bytes) -> None: ...
+    def on_message_begin(self) -> None: ...
+    def on_url(self, url: bytes) -> None: ...
+    def on_header(self, name: bytes, value: bytes) -> None: ...
+    def on_headers_complete(self) -> None: ...
+    def on_body(self, body: bytes) -> None: ...
+    def on_message_complete(self) -> None: ...
+    def on_chunk_header(self) -> None: ...
+    def on_chunk_complete(self) -> None: ...
+    def on_status(self, status: bytes) -> None: ...
```
Missing `self`.

Kludex commented on httptools/parser/url_parser.pyi:
```diff
@@ -10,18 +10,5 @@ class URL:
-def parse_url(url: Union[bytes, bytearray, memoryview, array]) -> URL:
-    """Parse URL strings into a structured Python object.
-
-    Returns an instance of ``httptools.URL`` class with the
-    following attributes:
-
-      - schema: bytes
-      - host: bytes
-      - port: int
-      - path: bytes
-      - query: bytes
-      - fragment: bytes
-      - userinfo: bytes
```
Irrelevant. The `URL` is defined above.

Kludex commented on httptools/parser/__init__.py:
```diff
@@ -1,6 +1,29 @@
 
-__all__ = parser.__all__ + errors.__all__ + url_parser.__all__  # NoQA
+__all__ = (
+    # protocol
+    "HTTPProtocol",
+    # parser
+    "HttpParser",
+    "HttpRequestParser",
+    "HttpResponseParser",
+    # errors
+    "HttpParserError",
+    "HttpParserCallbackError",
+    "HttpParserInvalidStatusError",
+    "HttpParserInvalidMethodError",
+    "HttpParserInvalidURLError",
+    "HttpParserUpgrade",
+    # url_parser
+    "parse_url",
+)
```
Proper definition.

Kludex committed:
Add py.typed

Kludex committed:
drop unneeded Union

Kludex commented:
I've dropped the `Union`s since they are not needed in `pyi` files.

Kludex committed:
move py.typed up

Kludex commented:
I've tested this on uvicorn, it works like a charm. It would be nice to add a type checker in this repository.

EDIT: I meant the typing... 😅

fantix commented:
Thank you! Looks there's an error: `ImportError: cannot import name 'HttpParser' from 'httptools.parser.parser'`

Let me add a CI workflow to check typing.

Kludex commented:
> Thank you! Looks there's an error: `ImportError: cannot import name 'HttpParser' from 'httptools.parser.parser'`
> 
> 
> 
> Let me add a CI workflow to check typing.

Then it seems the issue is that the `parser.pyi` exists? Should be `__init__.pyi`? 🤔

I'm on a plane. I'll fix this tomorrow.

Kludex committed:
Push CI

Kludex commented:
> Thank you! Looks there's an error: `ImportError: cannot import name 'HttpParser' from 'httptools.parser.parser'`
> 
> Let me add a CI workflow to check typing.

It seems `HttpParser` is marked as internal. I've fixed it.

I've added the type checker to the pipeline, but feel free to tweak whatever you want.

Kludex commented:
Btw, this gives less frustration to people when contributing to other projects:

<img width="865" height="403" alt="Screenshot 2025-10-20 at 10 27 44" src="https://github.com/user-attachments/assets/0baefffa-a38e-460b-b48b-02189e746ef3" />

You can find it on:

<img width="318" height="232" alt="Screenshot 2025-10-20 at 10 28 20" src="https://github.com/user-attachments/assets/715d9825-5c2b-43c0-977c-6208f28198c9" />

fantix committed:
Merge remote-tracking branch 'origin/master' into fix-typing

fantix committed:
Bump setup-python and use 3.9 for typechecking
